### PR TITLE
wildcard: add `useAnimatedAlert` hook

### PR DESCRIPTION
**Pull request by courier-new**
Recently while building the RBAC admin UI, we had a need for a toast-like alert: one which appears on the page after some action is taken to announce either the success of failure of that action, disappears automatically after a few seconds, but is clearly scoped to one item on one part of the page and thus appears in close proximity to it. Having items pop in and out of visibility is obviously bad UX in a number of regards, so we implemented a simple animation with `react-spring` to fade the alert in and grow its height as it appears, the fade it out and shrink the height once more as it disappears.

I realized after implementing this that there were several other places in the Batch Changes area that could benefit from such an animated, auto-dismissing alert, and thus the `useAnimatedAlert` hook was born. This hook simply provides a common interface for controlling the animation and auto-dismissing behavior of an alert, if such behavior is desired, as demonstrated in the new Storybook story section:

https://user-images.githubusercontent.com/8942601/227121360-724426b4-21ef-4903-81b0-efb88754ea62.mov

---

**The hook is screenreader-friendly.** As demonstrated in the Storybook story examples, screenreader announcements for the alert are enabled by passing a screenreader message to the hook, turning off `aria-live` for the actual `Alert` component (since it's always going to be rendered), and syncing `aria-hidden` of the `Alert` to the hook's `isShown` state. The hook will call `screenReaderAnnounce` with the provided message when the alert is shown.

**The hook is not officially ordained by design.** I kinda just made up values for the duration the alert lasts based on what felt right. 😅 The main goal was just something reusable for auto-dismissibility without jarring layout shift. In most cases, I expect auto-dismissibility is _not_ the best choice for alerts. And maybe there's a more toast-y alternative that we would eventually design for. If it doesn't feel right to incorporate this into Wildcard proper, I'm also happy to move it somewhere less visible where its usage wouldn't be as strongly encouraged.

## Test plan

Storybook story coverage. Tested with a screenreader. Manually tested the RBAC roles page and batch changes preview list for functionality.

## App preview:

- [Web](https://sg-web-kr-reusable-animated-alert-hook.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by courier-new** [comment]
@vovakulikov tagging you for the Wildcard part of this review. 🙇‍♀️ As mentioned in my PR description, if this feels too niche to actually be incorporated into Wildcard, I'm happy to move it somewhere else, too! 🙂

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.00% (-0.08 kb) | 0.02% (+2.57 kb) | 0.02% (+2.65 kb) | 0.13% (+1) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/kr/reusable-animated-alert-hook/compare-report.html;tab=live_object) for a full comparison between the commits 9a72533 and 4c1785b or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/wildcard/src/components/Alert/useAnimatedAlert.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        startTimer()
+    }, [startTimer, ariaAnnouncement.message, ariaAnnouncement.politeness])
+
+    const dismiss = useCallback(() => {
+        setShowAlert(false)
+        stopTimer()
+    }, [stopTimer])
+
+    const ref = useRef<HTMLDivElement>(null)
+    let height = ref.current?.offsetHeight || APPROX_MIN_BANNER_HEIGHT_PX
```
Comment (vovakulikov) [inline client/wildcard/src/components/Alert/useAnimatedAlert.ts:91]: Reading offsetHeight will trigger layout thrashing; probably we could use resize observable to run only when the element gets new size.

**Review by vovakulikov — APPROVED** [review]
Thanks for this @courier-new, I left one comment about size tracking, but overall looks good. I don't have big preferences about the right place for this, but probably batch changes would be better because 
- So far, it has only batch changes as consumer 
- I think we don't want to adopt this pattern widely at the moment because there was a short conversation about changing the idea behind toast UI for async actions response https://sourcegraph.slack.com/archives/C0HMGV90V/p1677180181407859 So maybe we will get similar to alert but toast UI 

but even with these points, I think we can have it in the wildcard, just let leave a note there that if you want to use it, ask someone from the design team about toast UI))

**Comment by vovakulikov** [comment]
@courier-new I would play with spring values in order to avoid visual overlapping between the content below and the disappearing alert 

Screenshot 2023-03-23 at 11 15 44

File: client/wildcard/src/components/Alert/useAnimatedAlert.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+}: AnimatedAlertOptions): AnimatedAlertControls => {
+    const [showAlert, setShowAlert] = useState<boolean>(defaultShown)
+
+    // Use a stopwatch to show the alert for a certain duration.
+    const {
+        time: { seconds },
+        start: startTimer,
+        stop: stopTimer,
+        isRunning,
+    } = useStopwatch(defaultShown)
```
Comment (vovakulikov) [inline client/wildcard/src/components/Alert/useAnimatedAlert.ts:69]: I think you can simplify the implementation of this a lot with useTransition hook. I think it was designed specifically for enter/exit animations https://www.react-spring.dev/docs/components/use-transition

**Comment by eseliger** [comment]
As part of #wg-delete-week I am going to close this PR. If it’s still needed, feel free to reopen and restore the branch.

**Pull request closed without merging**